src/new-components/tabs/tabs.tsx

```diff
@@ -0,0 +1,228 @@
+     */
+    space?: ResponsiveProp<Space>
+}
+
+/**
+ * A component used to group `<Tab>` elements together.
+ */
+function TabList({
+    children,
+    space = 'medium',
+    ...props
+}: TabListProps): React.ReactElement | null {
+    const tabContextValue = React.useContext(TabsContext)
+
+    if (!tabContextValue) {
+        return null
+    }
+
+    const { color, variant, ...tabState } = tabContextValue
+
+    return (
+        <BaseTabList {...props} {...tabState}>
+            <Inline space={space}>{children}</Inline>
+        </BaseTabList>
+    )
+}
+
+type TabPanelProps = {
+    /** The content to be rendered inside the tab */
+    children: React.ReactNode
+
+    /** The tabPanel's identifier. This must match its corresponding `<Tab>`'s id */
+    id: string
+
+    /**
+     * By default, the tab panel's content is always rendered even when they are not active. This behaviour can be changed to
+     * 'active', which renders only when the tab is active, and 'lazy', meaning while inactive tab panels will not be rendered
+     * initially, they will remain mounted once they are active until the entire Tabs tree is unmounted.
+     */
+    render?: 'always' | 'active' | 'lazy'
+}
+
+/**
+ * Used to define the content to be rendered when a tab is active. Each `<TabPanel>` must have a corresponding `<Tab>` component.
+ */
+function TabPanel({ children, id, render = 'always' }: TabPanelProps): React.ReactElement | null {
+    const tabContextValue = React.useContext(TabsContext)
+    const [tabRendered, setTabRendered] = React.useState(false)
+    const tabIsActive = tabContextValue?.selectedId === id
+
+    React.useEffect(
+        function trackTabRenderedState() {
+            if (!tabRendered && tabIsActive) {
+                setTabRendered(true)
+            }
+        },
+        [tabRendered, tabIsActive],
+    )
+
+    if (!tabContextValue) {
+        return null
+    }
+
+    const { color, variant, ...tabState } = tabContextValue
+
+    return (
+        <BaseTabPanel tabId={id} {...tabState}>
+            {render === 'always' ? children : null}
+            {render === 'active' && tabIsActive ? children : null}
+            {render === 'lazy' && (tabIsActive || tabRendered) ? children : null}
+        </BaseTabPanel>
+    )
+}
+
+type TabAwareSlotProps = {
+    /**
+     * Render prop used to provide the content to be rendered inside the slot. The render prop will be
+     * called with the current `selectedId`
+     */
+    children: (provided: { selectedId?: string | null }) => React.ReactElement | null
+}
+
+/**
+ * Allows content to be rendered based on the current tab being selected while outside of the TabPanel
+ * component. Can be placed freely within the main `<Tabs>` component.
+ */
+function TabAwareSlot({ children }: TabAwareSlotProps): React.ReactElement | null {
+    const tabContextValue = React.useContext(TabsContext)
+
+    return tabContextValue ? children({ selectedId: tabContextValue.selectedId }) : null
```

Comment:
I wonder if there's anything interesting in the `tabContextValue` to be shared with consumers of this component. No need to look into it now, though. We can wait until the need arises, if ever.

Reply:
Here is the full list: https://github.com/reakit/reakit/blob/a211d94da9f3b683182568a56479b91afb1b85ae/packages/reakit/src/Tab/TabState.ts#L13

I guess because the tab state isn't documented, they feel a bit internal to Reakit tab's implementation, although I've made use of `select` already. That said, I don't think there's much else in there that needs to be exposed for now, let's wait for a use case as you said 👍

Reply:
Yeah, it's ok for now. If something comes up, we'll see what to do about it.